src/server/pps/server/api_server_test.go

```diff
@@ -149,40 +150,84 @@ func TestCreateDatum(t *testing.T) {
+		require.True(t, stream.IsEOS(err))
+		require.Equal(t, remaining, n)
+	})
+	t.Run("CrossInput", func(t *testing.T) {
+		datumClient, err := pc.PpsAPIClient.CreateDatum(ctx)
+		require.NoError(t, err)
+		input := client.NewCrossInput(
+			client.NewPFSInput(pfs.DefaultProjectName, repo, "/file?"),
+			client.NewPFSInput(pfs.DefaultProjectName, repo, "/file?"),
+		)
+		require.NoError(t, datumClient.Send(&pps.CreateDatumRequest{Body: &pps.CreateDatumRequest_Start{Start: &pps.StartCreateDatumRequest{Input: input, Number: 50}}}))
+		n, err := grpcutil.Read[*pps.DatumInfo](datumClient, make([]*pps.DatumInfo, 50))
+		require.NoError(t, err)
+		require.Equal(t, 50, n)
+
+		require.NoError(t, datumClient.Send(&pps.CreateDatumRequest{Body: &pps.CreateDatumRequest_Continue{Continue: &pps.ContinueCreateDatumRequest{}}}))
+		n, err = grpcutil.Read[*pps.DatumInfo](datumClient, make([]*pps.DatumInfo, 51))
 		require.True(t, stream.IsEOS(err))
 		require.Equal(t, 50, n)
 	})
-	t.Run("UnimplementedInputTypes", func(t *testing.T) {
-		inputs := []*pps.Input{
-			{Union: []*pps.Input{}},
-			{Cross: []*pps.Input{}},
-			{Join: []*pps.Input{}},
-			{Group: []*pps.Input{}},
-		}
-		for _, input := range inputs {
-			datumClient, err := pc.PpsAPIClient.CreateDatum(ctx)
-			require.NoError(t, err)
-			require.NoError(t, datumClient.Send(&pps.CreateDatumRequest{Body: &pps.CreateDatumRequest_Start{Start: &pps.StartCreateDatumRequest{Input: input}}}))
-			_, err = datumClient.Recv()
-			require.ErrorContains(t, err, "unimplemented input type")
-		}
+	t.Run("JoinInput", func(t *testing.T) {
+		datumClient, err := pc.PpsAPIClient.CreateDatum(ctx)
+		require.NoError(t, err)
+		input := client.NewJoinInput(
+			client.NewPFSInputOpts(repo, pfs.DefaultProjectName, repo, "master", "/file(?)", "$1", "", false, false, nil),
+			client.NewPFSInputOpts(repo, pfs.DefaultProjectName, repo, "master", "/file(?)", "$1", "", false, false, nil),
+		)
+		require.NoError(t, datumClient.Send(&pps.CreateDatumRequest{Body: &pps.CreateDatumRequest_Start{Start: &pps.StartCreateDatumRequest{Input: input}}}))
+		n, err := grpcutil.Read[*pps.DatumInfo](datumClient, make([]*pps.DatumInfo, 11))
+		require.True(t, stream.IsEOS(err))
+		require.Equal(t, 10, n)
+	})
+	t.Run("GroupInput", func(t *testing.T) {
+		datumClient, err := pc.PpsAPIClient.CreateDatum(ctx)
+		require.NoError(t, err)
+		input := client.NewGroupInput(
+			client.NewPFSInputOpts(repo, pfs.DefaultProjectName, repo, "master", "/file(?)*", "", "$1", false, false, nil),
+			client.NewPFSInputOpts(repo, pfs.DefaultProjectName, repo, "master", "/file(?)*", "", "$1", false, false, nil),
+		)
+		require.NoError(t, datumClient.Send(&pps.CreateDatumRequest{Body: &pps.CreateDatumRequest_Start{Start: &pps.StartCreateDatumRequest{Input: input}}}))
+		n, err := grpcutil.Read[*pps.DatumInfo](datumClient, make([]*pps.DatumInfo, 11))
+		require.True(t, stream.IsEOS(err))
+		require.Equal(t, 10, n)
```

Comment:
Somewhere, there needs to be a test that actually checks that CreateDatum produces correct result.  You run all the code in these tests, which is good, but you aren't checking that the results are correct.  Anything that returns 10 copies of DatumInfo (perhaps nil 10 times) will pass the test, which is not good enough for someone to be able to work on this code confidently.

Reply:
I tested correctness for the inputs manually. I'll add some of those to the test suite. 👍